ovrlib/pa: client for PA

liewegas opened the pull request.

liewegas committed:
ovrlib/pa: initial version of PA OVR client

liewegas committed:
.gitignore: *~, *.swap (emacs, vi)

liewegas committed:
factor out address vs unit normalization

liewegas committed:
improve request

liewegas committed:
fix post

liewegas committed:
put constants in source

save ourselves several API calls fetching static information

liewegas committed:
asdf

liewegas committed:
fix request xml/json

liewegas committed:
remove special county handling

API takes the county name, not the id

liewegas committed:
fix signature mode

liewegas committed:
fix gender. clean up constants

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+        root = etree.fromstring(xml_str)
+
+        # check for errors only
+        if root.tag == "RESPONSE":
```
We were seeing that incomplete requests would return something without an error, but also no registration ID, right? Should we throw an error for that case?

nickcatal commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+
+        print("UNIT_TYPE = %s\n" % json.dumps(unit_type, indent=4))
+        print("PARTY = %s\n" % json.dumps(party, indent=4))
+        print("GENDER = %s\n" % json.dumps(gender, indent=4))
```
`print` here and elsewhere

Probably not the world's worst idea to use `logger.debug` here instead of removing the print outright

nickcatal commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+    def register(self, registration, signature=None, signature_type=None, is_new=True):
+
+        """
+        Submit a voter registratio
```
missing a `n`

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+
+        return root
+
+    def fetch_constants(self):
```
Should we return some data structure here rather than printing? We can have a little CLI in a separate file to print out the data, but it'd be nice for this to return a machine-readable data structure if someone wants to, e.g., do validation or render UI dropdowns based on this.

nickcatal commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+        a recognized unit is found in the address line(s), it will be moved into the
+        dedicated unittype+unitnumber fields.
+        """
+        if addr.get("unittype") is not None or addr.get("unitnumber") is not None:
```
Do you ever see `unittype` or `unitnumber` ever being a falsy value?

If not, I think this can just be `if addr.get("unittype") or addr.get("unitnumber")`

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+
+        vals = {
+            "batch": "0",
+            "united-states-citizen": "1",
```
I'm a little nervous about always setting these to "1" -- I think forcing people to pass in `will_be_eighteen_on_election_day: True, us_citizen: True, made_declarations: True` and then erroring if those values aren't passed is a good way to reinforce that those are things you have to render in the UI and have the user agree to.

benweissmann replied:
As a bonus, if we define like `to_request_body()` on the dataclass, then library users who don't want to use this library for the request logic (and just want to use it for the PITA part of assembling a valid request) can just import the dataclass, instantiate it, and call `data.to_request_body()` to get a valid request body.

nickcatal commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+
+        # first try address2 field
+        if "address2" in addr:
+            m = re.compile(r"^#?(\d+)$").match(addr["address2"].strip())
```
We probably want to do all these `re.compile` steps when python first loads the module (thus have them above the class) then do just the matching using the pre-compiled regex inside the class.

Regex compiling can be kinda intense and doesn't need to be done on a per-registration basis, so we can front-load it to when python first launches.

benweissmann commented on ovrlib/tests/test_pa.py:
```diff
@@ -0,0 +1,39 @@
+from ..pa import Session
+
+
+@pytest.mark.parametrize(
```
I think it'd also be useful to have one or two end-to-end tests using [responses](https://github.com/getsentry/responses) would be valuable here

benweissmann replied:
Also a test of `fetch_constants`

nickcatal commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+XML_TEMPLATE = "<APIOnlineApplicationData xmlns='OVRexternaldata'>  <record>    <batch></batch>    <FirstName></FirstName>    <MiddleName></MiddleName>    <LastName></LastName>    <TitleSuffix></TitleSuffix>    <united-states-citizen></united-states-citizen>    <eighteen-on-election-day></eighteen-on-election-day>    <isnewregistration></isnewregistration>    <name-update></name-update>    <addre ...
+
+
+class InvalidAccessKeyError(Exception):
```
Do we want all of these to inherit from a master `OVRLibException`? We also may want these in a separate `exceptions.py` file

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+
+"""
+
+UNIT_TYPE = {
```
Are these constants from the PDF? Or coming from the API? Would be good to document where these came from and how to update them.

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+    pass
+
+
+class ReadOnlyAccessKeyError(Exception):
```
Is this thrown somewhere? I definitely think surfacing this error in particular with some message like "please email your point-of-contact at the Pennsylvania Secretary of State's office to upgrade your API key to read-write access" would be good, given that this is an issue we ran into.

benweissmann commented on ovrlib/pa.py:
```diff
@@ -0,0 +1,549 @@
+                else:
+                    raise InvalidRegistrationError(f"gender '{v}' not recognized; must be one of {GENDER}")
+            elif k == "suffix":
+                # NOTE: the API enumerates specific suffixes; not sure what happens if we
```
Can we test this out?

benweissmann reviewed (changes requested):
This looks great! I have a bunch of comments mostly around ways that I think we can make this more useful to other groups.

I also think that having a really simple CLI (in a separate file) would be helpful in terms of providing something that people can use out-of-the-box to test their API key, etc.

nickcatal reviewed (commented):
Building on top of @benweissmann's comments, here are my much more mundane style comments

liewegas committed:
API handles any suffix

liewegas committed:
address most review comments

liewegas committed:
add test-pa.py CLI tool

liewegas committed:
simplify federal voter

liewegas committed:
fix units

liewegas committed:
add some tests

liewegas committed:
separate exceptions.py

benweissmann reviewed (changes requested):
Sorry, I had one more larger comment from before that it looks like I didn't actually hit save on:

I think we should use [MyPy](http://mypy-lang.org/) here and use [dataclasses](https://docs.python.org/3/library/dataclasses.html) instead of dicts for our inputs/outputs. I think this will help with a couple of things when other folks go to use this library:

- We're going to need to document the structure of the input/output dictionaries somewhere, so we might as well do that in a way that mypy can type-check (I think sphinx can auto-generate good docs based on type hints / dataclasses)

- If whoever's using this library is also using mypy, they'll get nice compile-time errors if they mess up any of the data structures

As another plus, I think that can help break this up a bit -- we can create three dataclasses in separate files: `PAOVRRequest`, `PAOVRResponse`, and `PAOVRConstants` that will enumerate all the required/option fields, and we can implement `to_request_body()` on `PAOVRRequest`, `from_response_body()` on `PAOVRResponse`, and `from_response_bodies()` on `PAOVRConstants`.

If you haven't used mypy before, it can have a bit of a learning curve (the [vscode python extension](https://code.visualstudio.com/docs/python/linting) helps a lot), so lmk if you want to pair! There's some examples of where we've used them to document and validate complex data structures in the [fax gateway](https://github.com/vote/fax_gateway/blob/master/app/helpers/messages.py) and [Turnout](https://github.com/vote/turnout/blob/ab6bcaca9f8d448606964085112451c37c17e90c/app/common/pdf/pdftemplate.py#L16).

liewegas committed:
common parent class for exceptions

liewegas committed:
use dataclass for registration and response

liewegas committed:
add mypy, fix tests

liewegas commented:
Moved to dataclasses, except for the constants: dataclasses don't like default values that are mutable dicts/lists/etc.

Also added mypy and type notations

liewegas committed:
separate out election info into separate method

benweissmann reviewed (approved):
Looks great! Methods just need return type annotations (e.g. `get_election_info(self) -> PAOVRElectionInfo:`)

liewegas committed:
more type annotations, and a few resulting fixes

liewegas committed:
add setup.py and __version__.py

liewegas committed:
pip freeze
